### DISCUSSION TITLE
What does it mean to expose a schema in the API?

### DISCUSSION THREAD

**Discussion by vincenzon** [supabase/supabase#21096] [Questions] [answered]
I am trying to access functions in a private schema using rpc calls in an edge function. The supabase client in the edge function has the service role key, so I assumed it would be able to access a private schema. However, I get an error that only public, storage, and graphql_public are allowed. I see a settings option to "expose" a schema in the API:

```
The schemas to expose in your API. Tables, views and stored procedures in these schemas will get API endpoints.public and storage are protected by default.
```

But it seems wrong to expose a private schema. Does this mean functions in this schema would become accessible to users with the anon key? What about logged in users? 

Can someone clarify what the security implications are of exposing a schema in the API are?

Thank you.

**Comment by GaryAustin1** [comment]
If you want to make an rpc call to a custom schema then you have to make it available to the API.  

With public schema you can block functions individually by removing grants to public,anon and authenticated roles.  With a custom schema you can control grants to all functions in the schema by setting roles up front that will have access.

This shows setting up grants on a custom schema:  https://supabase.com/docs/guides/api/using-custom-schemas
To revoke grants for a specific function in the public schema:
```sql
REVOKE EXECUTE ON FUNCTION restricted_function() FROM PUBLIC;
REVOKE EXECUTE ON FUNCTION restricted_function() FROM anon;
REVOKE EXECUTE ON FUNCTION restricted_function() FROM authenticated;
```

Service_role only bypass RLS.  It has no other powers as far as the database goes.

**Reply by vincenzon** [reply]
Thanks Gary. Follow up question, before I saw your answer, I was going to move the table that is the subject of my rpc calls into the public schema, activate RLS, but have no policies for reading / writing to it. Would that be roughly equivalent to putting the table into the private schema but having the rpc calls in the public schema and revoking access as you described? Thanks for your help.

**Reply by GaryAustin1** [reply, marked as answer]
Either can work.
Locking the tables is certainly required if they are in the public schema.  Your rpc called function can read them either in public or private though even if it is in public schema.

If your function does nothing but access the tables then RLS should be sufficient as anon and authenticated can't do anything.  If your function does anything else then you might need to block access.  Or you can do both.
